packages/panzer/disc-fe/src/Panzer_L2Projection_impl.hpp

```diff
@@ -95,55 +95,56 @@ namespace panzer {
       // Based on descriptor, currently assumes there should only be one workset
       panzer::WorksetDescriptor wd(block,panzer::WorksetSizeType::ALL_ELEMENTS,true,true);
       const auto& worksets = worksetContainer_->getWorksets(wd);
-      TEUCHOS_ASSERT(worksets->size() == 1);
-      const auto& workset = (*worksets)[0];
-      const auto& basisValues = workset.getBasisValues(targetBasisDescriptor_,integrationDescriptor_);
 
-      const auto& unweightedBasis = basisValues.basis_scalar;
-      const auto& weightedBasis = basisValues.weighted_basis_scalar;
-
-      // Offsets (this assumes UVM, need to fix)
-      const std::vector<LO>& offsets = targetGlobalIndexer_->getGIDFieldOffsets(block,fieldIndex);
-      PHX::View<LO*> kOffsets("MassMatrix: Offsets",offsets.size());
-      for(const auto& i : offsets)
-        kOffsets(i) = offsets[i];
-
-      PHX::Device::fence();
-
-      // Local Ids
-      Kokkos::View<LO**,PHX::Device> localIds("MassMatrix: LocalIds", workset.numOwnedCells()+workset.numGhostCells()+workset.numVirtualCells(),
-                                              targetGlobalIndexer_->getElementBlockGIDCount(block));
-
-      // Remove the ghosted cell ids or the call to getElementLocalIds will spill array bounds
-      const auto cellLocalIdsNoGhost = Kokkos::subview(workset.cell_local_ids_k,std::make_pair(0,workset.numOwnedCells()));
-
-      targetGlobalIndexer_->getElementLIDs(cellLocalIdsNoGhost,localIds);
-
-      const int numBasisPoints = static_cast<int>(weightedBasis.extent(1));
-      Kokkos::parallel_for(workset.numOwnedCells(),KOKKOS_LAMBDA (const int& cell) {
-
-        LO cLIDs[256];
-        const int numIds = static_cast<int>(localIds.extent(1));
-        for(int i=0;i<numIds;++i)
-          cLIDs[i] = localIds(cell,i);
-
-        double vals[256];
-        const int numQP = static_cast<int>(unweightedBasis.extent(2));
-
-        for (int qp=0; qp < numQP; ++qp) {
-          for (int row=0; row < numBasisPoints; ++row) {
-            int offset = kOffsets(row);
-            LO lid = localIds(cell,offset);
-
-            for (int col=0; col < numIds; ++col)
-              vals[col] = unweightedBasis(cell,row,qp) * weightedBasis(cell,col,qp);
+      for (const auto& workset : *worksets) {
 
-            M.sumIntoValues(lid,cLIDs,numIds,vals,true,true);
+        const auto& basisValues = workset.getBasisValues(targetBasisDescriptor_,integrationDescriptor_);
+        
+        const auto& unweightedBasis = basisValues.basis_scalar;
+        const auto& weightedBasis = basisValues.weighted_basis_scalar;
+        
+        // Offsets (this assumes UVM, need to fix)
```

Comment:
Is there an issue to capture the need to fix this at some later point?

Reply:
Nope.  Will need to add a general ticket to remove UVM use in Pazer. Panzer assumes UVM in a ton of places. As I go over code, I have been adding this comment to flag sections as a reminder. Removing all UVM will be a significant undertaking. Note this code is not actually new. I wrapped this whole function in a new loop over worksets, resulting in an indentation change. The diff tool couldn't figure this out and flagged the whole section as changed.

Reply:
Gotcha.  Sounds good to me.